### PR TITLE
Fix diagnostic when a variable conflicts with an enclosing lambda parameter.

### LINKED ISSUE #12052: A misleading error is reported for a local or parameter conflicting with a range variable from an outer scope
```
        var res = from a in new[] { 1 }
                  select (Func<int, int>)(a => 1);
```

Observed:
                  select (Func<int, int>)(**a** => 1);
error CS0412: 'a': a parameter or local variable cannot have the same name as a method type parameter

Expected:
An error indicating a conflict with a range variable.

This is going to affect pattern variable declarations and out variable declarations.

### PR DESCRIPTION
**Pull request by gafter**
Fixes #12052

@dotnet/roslyn-compiler Please review this tiny diagnostic correction.

### CODE REVIEW

**Comment by cston** [comment]
LGTM

**Comment by gafter** [comment]
@dotnet-bot Please test windows_debug_vs-integration_prtest

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameCollisionTests.cs [hunk 1]
```diff
@@ -828,31 +828,30 @@ class Class
...
-    //                     int name3 = 2;                   // 0136
-    Diagnostic(ErrorCode.ERR_LocalSameNameAsTypeParam, "name3").WithArguments("name3").WithLocation(27, 25),
-    // (32,17): error CS0136: A local or parameter named 'name1' cannot be declared in this scope because that name is used in an enclosing local scope to define a local or parameter
-    //         return (name1, name2) => name1;              // 0136 on both name1 and name2
-    Diagnostic(ErrorCode.ERR_LocalIllegallyOverrides, "name1").WithArguments("name1").WithLocation(32, 17),
-    // (32,24): error CS0136: A local or parameter named 'name2' cannot be declared in this scope because that name is used in an enclosing local scope to define a local or parameter
-    //         return (name1, name2) => name1;              // 0136 on both name1 and name2
-    Diagnostic(ErrorCode.ERR_LocalIllegallyOverrides, "name2").WithArguments("name2").WithLocation(32, 24)
-);
+                //         foreach (var name1 in "string")            // 0136
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NameCollisionTests.cs]: >//         foreach (var name1 in "string")            // 0136 [](start = 16, length = 61)

Missing error text for this location.

**Comment by AlekseyTs** [comment]
Done with review pass.

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a second review?

**Pull request merged**